plugin/src/main/kotlin/com/nishtahir/CargoBuildTask.kt

```diff
@@ -162,3 +176,86 @@ open class CargoBuildTask : DefaultTask() {
         }.assertNormalExitValue()
     }
 }
+
+// We don't want to invoke `rustc -Vv` a ton, but we can't invoke it at all without
+// the gradle project (despite this not being dependent on the project). As a result,
+// we need to do some locking in order to cache the result of that.
+//
+// This object exists to encapsulate that, it just exposes a `fun getValue(p: Project): String?`.
+//
+// TODO: Is there somewhere more natural to put configuration state like this?
+private object DefaultTargetTriple {
+    // Initialized once, write-protected by `writeLock`, but no lock needs to be held when reading.
+    @Volatile
+    private var knownValue: String? = null
+
+    // Protects `knownValue` during writing. No protection is needed during reading. We also hold it
+    // while computing the value, to avoid wasting work.
+    private val writeLock = Object()
+
+    /**
+     * If we already know the triple, return it. Otherwise, call out to rustc to get it.
+     *
+     * On macOS, this should output `x86_64-apple-darwin`, for example. Returns null on any problems.
+     *
+     * This requires the `project` since I'm not sure how else to execute commands, but
+     * isn't actually dependent on the project.
+     */
+    internal fun getValue(project: Project): String? {
+
+        // Avoid synchronizing if we've initialized already.
+        //
+        // Note that this is not perfect, as `computeDefaultTargetTriple()` can return null in error
+        // cases, which will mean we'll compute the result multiple times. That said, this shouldn't
+        // happen in practice, and if it does, hopefully the log spam will push the user to file a
+        // rust-android-gradle bug.
+        knownValue?.let { cached ->
+            return cached
+        }
+        synchronized(writeLock) {
+            // Did anybody compute this while we were waiting?
+            knownValue?.let { value ->
+                return value
+            }
+            return computeDefaultTargetTriple(project)?.let { computed ->
+                // Note: We only write to knownValue if `computed` is not null.
+                //
+                // This isn't strictly necessary, but it's a good practice not
+                // to write to values that are read without synchronization.
+                knownValue = computed
+                computed
+            }
+        }
+    }
+
+    /** Attempt to get the default (desktop) target triple by parsing `rustc -Vv` output. */
+    private fun computeDefaultTargetTriple(project: Project): String? {
+        val stdout = ByteArrayOutputStream()
+        val result = project.exec { spec ->
```

Comment:
You're free to use Kotlin's Java-wrapping API, like https://github.com/mozilla-mobile/android-automation-tools/blob/32af3b4dfd1e5abb4d1b71de983102c12e8c7cbe/gradle-plugin/src/main/kotlin/org/mozilla/android/ext/Runtime.kt.  That seems better than `project.exec`.  If this fails, an `Exception` is better than logging and continuing.

Reply:
I disagree about an exception being better -- If we fail to get the default target triple we can just specify it explicitly and move on. If it really indicates a problem, then cargo will fail to locate rustc as well.

Doing that seems far more likely to cause a fire drill to me :p

Reply:
That said I did change this to use the Runtime.exec api, which let me replace manual locking with `Lazy<T>`